cEP-0016.md

```diff
@@ -0,0 +1,382 @@
+     tries to merge master or any other branch onto the pull request, message
+     is displayed to perform a `git rebase` instead.
+7. For revert commits:
+   - A meta bear called `RevertCommitBear` would be created.
+   - Revert commits are identified based on the format of the commit inside
+     `GitCommitBear`.
+   - For every revert commit the `RevertCommitBear` would return a tuple of
+     reverted commit hash and the reverting commit hash.
+   - Now using `AnalyzeCommitBear`, the patches of both commits are compared to
+     ensure that they are in fact reverse of each other. Minor differences are
+     allowed.
+   - In case of total mismatch, a note in the commit message highlighting that
+     it is not a clean revert is expected.
+   - A message suggesting improvements in the commit message is displayed.
+
+## Code Prototypes
+
+A prototype for identifying Travis merge commits:
+
+```python
+    if len(shortlog) > shortlog_length:
+        travis_merge_commit_regex = re.compile(r'^Merge [0-9a-f]+ into [0-9a-f]+$|....')
+
+        match = re.fullmatch(travis_merge_commit_regex, shortlog)
+        if match:
+            # ignore this commit from any further inspection
+            # inspect the unmerged parent of this commit instead
+```
+
+A prototype of `SkipCIBear` :
+
+```python
+   import abc
+   import re
+
+   from bears.vcs.CommitBear import _CommitBear
+   from coala_utils.ContextManagers import change_directory
+   from coalib.misc.Shell import run_shell_command
+   from coalib.settings.FunctionMetadata import FunctionMetadata
+
+
+   class SkipCIBear(CommitBear):
+    __metaclass__ = abc.ABCMeta
+
+    BEAR_DEPS = {'_CommitBear'}
+    LANGUAGES = {'Git'}
+    CAN_DETECT = {'Commits that disable CI build'}
+
+    CI_PROVIDERS_SUPPORTING_SKIPPING_BUILD = {
+        'circleci': (r'[skip ci]?'
+                     r'|[ci skip]?'),
+        'travisci': (r'[skip ci]?'
+                     r'|[ci skip]?'),
+        'semaphoreci': (r'[skip ci]?'
+                     r'|[ci skip]?'),
+        'gitlabci': (r'[skip ci]?'
+                     r'|[ci skip]?'),
+        'appveyor': None,
+    }
+
+    @classmethod
+    def get_ciskip_metadata(cls):
+        return FunctionMetadata.from_function(
+            cls.check_skip_ci_build,
+            omit={'self', 'commit_message'})
+
+    @classmethod
+    def get_CIengine_from_CIproviders(cls):
+        """
+        Retrieve first CI engine from the list of CI providers.
+        """
+
+    def run(self,
+            allow_skipping_ci_build: bool = False,
+            **kwargs):
+        """
+        This bear ensures that if the HEAD commit contains `[ci skip]` or
+        `[skip ci]` then it must not modify files in tests or 'coafile'.
+
+        :param allow_skipping_ci_build: Whether skipping ci build is
+                                        allowed by the organization or not.
+        """
+        commit_message = # get commit message at HEAD from _CommitBear
```

Comment:
I'm not sure, but this design looks like it will get stuck in circular dependency problems
I might be wrong on this, but `_CommitBear` is an abstract class which has functions that needs to be implemented by it's sub-classes, for e.g say a GitCommitBear (get_head_commit() will not work until a subclass provides a concrete implementation for it)

And since the GitCommitBear will have `SkipCIBear` as a bear_deps , `SkipCIBear` cannot have `GitCommitBear` as a bear_deps
otherwise it might go into circular dependency issues

(Same for other bears as well)

Reply:
I think this is the problem I am facing in this PR https://github.com/coala/coala-bears/pull/2523 The problem I am facing is fetching commits require that the vcs host is known and also these bears are common to all the git hosting providers. Any suggestions for better design?

Reply:
The functions are intended to be separate for GitCommitBear and HgCommitBear, since they are specific to `git` or `hg`

The first solution that comes to my mind to resolve these issues would be to keep these bears classes abstract, and do a multiple inheritance for say a GitSkipCIBear
e.g

`GitSkipCIBear(GitCommitBear, SkipCommitBear)`

This way you can provide an implementation without worrying about VCS specifics

Reply:
What would `SkipCommitBear` be like?

Reply:
Looking at the PR, I think the best way would be to define the `check_skip_ci_build` function inside _CommitBear itself and leave the VCS specific tasks to the sub-classes (thereby eliminating the need for a SkipCIBear)

Can you elaborate a little on why did we need a SkipCIBear separately ?

Reply:
>  I think the best way would be to define the check_skip_ci_build function inside _CommitBear itself

Even I feel the same. 

> Can you elaborate a little on why did we need a SkipCIBear separately ?

I went for a separate bear because the corresponding issue has tag "bear-proposal". Maybe @jayvdb can throw some light on that.